### PR TITLE
[BEAM-8902] parameterize input type of Java external transform

### PR DESCRIPTION
**Pull request by ihji**
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by ihji** [comment]
R: @robertwb

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java [hunk 1]
```diff
@@ -139,7 +139,7 @@ public void expandPythonTest() {
           testPipeline
               .apply(Create.of("1", "2", "2", "3", "3", "3"))
               .apply(
-                  External.<KV<String, Integer>>of(
+                  External.<PCollection<String>, KV<String, Integer>>of(
```
Comment (robertwb) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java]: From a usability point of view, is it possible to infer this first type argument (as it must match the input type)? Presumably we were inferring String before...

Comment (ihji) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java]: It's generally impossible to only infer the first type parameter when the second type parameter needs to be explicitly defined. There's a workaround however. I updated the example.

**Comment by ihji** [comment]
Run Java PreCommit

**Comment by ihji** [comment]
Run Python2_PVR_Flink PreCommit

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java (hunk 1 above)
Comment (robertwb) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java]: Hmm... That's unfortunate.

I wonder if we should have separate methods for the common "applies to a single collection" and more generic case.

Comment (ihji) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java]: Yeah, we can create a variant of `of()` for the common case. Do you have any good naming suggestion? :)

**Comment by ihji** [comment]
added `withOutputType()` which users can use to specify the element type of the output pcollection.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java [hunk 2]
```diff
@@ -139,8 +139,8 @@ public void expandPythonTest() {
           testPipeline
               .apply(Create.of("1", "2", "2", "3", "3", "3"))
               .apply(
-                  External.<KV<String, Integer>>of(
-                      "beam:transforms:xlang:count", new byte[] {}, target))
+                  External.of("beam:transforms:xlang:count", new byte[] {}, target)
```
Comment (robertwb) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ExternalTest.java]: This still feels a bit odd, but it's good enough that I don't want to block the testing work. Could you please start a discussion on the dev list about what the best API should be (for {single,multiple}{input,output}?)

**Pull request merged**